File: 45 Jump Game II/solution.py
# -*- coding: utf-8 -*-


class Solution(object):

    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        jumps = 0
        curMax = 0
        curReach = 0
        for i in range(len(nums)):
            if curReach < i:
                jumps += 1
                curReach = curMax
            curMax = max(nums[i] + i, curMax)
        return jumps


checker = Solution()
test = [2, 3, 1, 1, 4]
print checker.jump(test)

File: 45 Jump Game II/solution.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int jumps = 0;
        int curReach = 0;
        int curMax = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (curReach < i) {
                jumps++;
                curReach = curMax;
            }    
            curMax = max(curMax, nums[i] + i);
        }
        return jumps;
    }
};
